File: PERL/flynn.pl
use strict;
use warnings;
use JSON;
use IO::Prompter;
use IO::Socket::SSL;
use Time::HiRes qw(sleep);
use File::Slurp;

my %cl = (
    gr => "\x1b[32m",
    gb => "\x1b[4m",
    br => "\x1b[34m",
    st => "\x1b[9m",
    yl => "\x1b[33m",
    rt => "\x1b[0m"
);

my $socket = undef;
my $ping_interval;
my $countdown_interval;
my $potential_points = 0;
my $countdown = "Calculating...";
my $points_total = 0;
my $points_today = 0;
my $reconnect_attempts = 0;
my $max_reconnect_attempts = 5;
my $max_reconnect_interval = 5 * 60; # 5 minutes in seconds
my $CoderMarkPrinted = 0;

sub CoderMark {
    return if $CoderMarkPrinted;
    print "\n╭━━━╮╱╱╱╱╱╱╱╱╱╱╱╱╱╭━━━┳╮
┃╭━━╯╱╱╱╱╱╱╱╱╱╱╱╱╱┃╭━━┫┃$cl{gr}
┃╰━━┳╮╭┳━┳━━┳━━┳━╮┃╰━━┫┃╭╮╱╭┳━╮╭━╮
┃╭━━┫┃┃┃╭┫╭╮┃╭╮┃╭╮┫╭━━┫┃┃┃╱┃┃╭╮┫╭╮╮$cl{br}
┃┃╱╱┃╰╯┃┃┃╰╯┃╰╯┃┃┃┃┃╱╱┃╰┫╰━╯┃┃┃┃┃┃┃
╰╯╱╱╰━━┻╯╰━╮┣━━┻╯╰┻╯╱╱╰━┻━╮╭┻╯╰┻╯╰╯$cl{rt}
╱╱╱╱╱╱╱╱╱╱╱┃┃╱╱╱╱╱╱╱╱╱╱╱╭━╯┃
╱╱╱╱╱╱╱╱╱╱╱╰╯╱╱╱╱╱╱╱╱╱╱╱╰━━╯
\n$cl{gb}Teneo Node Cli $cl{gr}v1.1.0 $cl{rt}$cl{gb}$cl{br}dev_build$cl{rt}\n";
    $CoderMarkPrinted = 1;
}

sub read_file {
    my ($file) = @_;
    return decode_json(read_file($file)) if -e $file;
    return {};
}

sub write_file {
    my ($file, $data) = @_;
    write_file($file, encode_json($data));
}

sub get_local_storage {
    return read_file('localStorage.json');
}

sub set_local_storage {
    my ($data) = @_;
    my $current_data = get_local_storage();
    my %new_data = (%$current_data, %$data);
    write_file('localStorage.json', \%new_data);
}

sub get_user_id_from_file {
    my $data = read_file('UserId.json');
    return $data->{userId} if $data->{userId};
    return undef;
}

sub set_user_id_to_file {
    my ($user_id) = @_;
    write_file('UserId.json', { userId => $user_id });
}

sub get_account_data {
    return read_file('DataAccount.json');
}

sub set_account_data {
    my ($email, $password, $access_token, $refresh_token, $personal_code) = @_;
    my %account_data = (
        email => $email,
        password => $password,
        access_token => $access_token,
        refresh_token => $refresh_token,
        personalCode => $personal_code
    );
    write_file('DataAccount.json', \%account_data);
}

sub get_reconnect_delay {
    my ($attempt) = @_;
    my $base_delay = 5; # 5 seconds
    my $additional_delay = $attempt * 5; # Additional 5 seconds for each attempt
    return $base_delay + $additional_delay < $max_reconnect_interval ? $base_delay + $additional_delay : $max_reconnect_interval;
}

sub connect_websocket {
    my ($user_id) = @_;
    return if $socket;

    my $version = "v0.2";
    my $url = "wss://secure.ws.teneo.pro/websocket?userId=" . uri_escape($user_id) . "&version=" . uri_escape($version);
    $socket = IO::Socket::SSL->new(PeerAddr => $url) or die "Could not connect to WebSocket: $!";

    # Handle WebSocket events
    # Note: WebSocket handling in Perl is more complex and may require additional libraries
}

sub disconnect_websocket {
    if ($socket) {
        close($socket);
        $socket = undef;
        stop_pinging();
    }
}

sub start_pinging {
    stop_pinging();
    $ping_interval = AnyEvent->timer(0, 10, sub {
        if ($socket) {
            my $ping_message = encode_json({ type => "PING" });
            print $socket $ping_message;
            set_local_storage({ lastPingDate => time() });
        }
    });
}

sub stop_pinging {
    if ($ping_interval) {
        undef $ping_interval;
    }
}

sub signal_handler {
    print 'Received SIGINT. Stopping pinging...';
    stop_pinging();
    disconnect_websocket();
    exit(0);
}

$SIG{INT} = \&signal_handler;

sub start_countdown_and_points {
    stop_countdown();
    update_countdown_and_points();
    $countdown_interval = AnyEvent->timer(0, 1, \&update_countdown_and_points);
}

sub update_countdown_and_points {
    my $local_storage = get_local_storage();
    if ($local_storage->{lastUpdated}) {
        my $next_heartbeat = $local_storage->{lastUpdated} + 15 * 60; # 15 minutes
        my $now = time();
        my $diff = $next_heartbeat - $now;

        if ($diff > 0) {
            my $minutes = int($diff / 60);
            my $seconds = $diff % 60;
            $countdown = "${minutes}m ${seconds}s";

            my $max_points = 25;
            my $time_elapsed = $now - $local_storage->{lastUpdated};
            my $time_elapsed_minutes = $time_elapsed / 60;
            my $new_points = $max_points < ($time_elapsed_minutes / 15) * $max_points ? $max_points : ($time_elapsed_minutes / 15) * $max_points;
            $new_points = sprintf("%.2f", $new_points);

            if (rand() < 0.1) {
                my $bonus = rand() * 2;
                $new_points = $max_points < ($new_points + $bonus) ? $max_points : ($new_points + $bonus);
                $new_points = sprintf("%.2f", $new_points);
            }

            $potential_points = $new_points;
        } else {
            $countdown = "Calculating...";
            $potential_points = 25;
        }
    } else {
        $countdown = "Calculating...";
        $potential_points = 0;
    }
    set_local_storage({ potentialPoints => $potential_points, countdown => $countdown });
}

sub get_user_id {
    my $login_url = "https://node-community-api.teneo.pro/auth/v1/token?grant_type=password";
    my $authorization = "Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6Imlra25uZ3JneHV4Z2pocGxicGV5Iiwicm9sZSI6ImFub24iLCJpYXQiOjE3MjU0MzgxNTAsImV4cCI6MjA0MTAxNDE1MH0.DRAvf8nH1ojnJBc3rD_Nw6t1AV8X_g6gmY_HByG2Mag";
    my $apikey = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6Imlra25uZ3JneHV4Z2pocGxicGV5Iiwicm9sZSI6ImFub24iLCJpYXQiOjE3MjU0MzgxNTAsImV4cCI6MjA0MTAxNDE1MH0.DRAvf8nH1ojnJBc3rD_Nw6t1AV8X_g6gmY_HByG2Mag";

    my $account_data = get_account_data();
    my $email = $account_data->{email} || prompt('Email: ');
    my $password = $account_data->{password} || prompt('Password: ');

    my $response = HTTP::Tiny->new->post($login_url, {
        headers => {
            'Authorization' => $authorization,
            'apikey' => $apikey
        },
        content => encode_json({ email => $email, password => $password })
    });

    if ($response->{success}) {
        my $data = decode_json($response->{content});
        my $access_token = $data->{access_token};
        my $refresh_token = $data->{refresh_token};
        print "Access_Token: $access_token\n";
        print "Refresh_Token: $refresh_token\n";

        my $AuthUserUrl = "https://node-community-api.teneo.pro/auth/v1/user";
        my $AuthResponse = HTTP::Tiny->new->get($AuthUserUrl, {
            headers => {
                'Authorization' => "Bearer $access_token",
                'apikey' => $apikey
            }
        });

        if ($AuthResponse->{success}) {
            my $user_data = decode_json($AuthResponse->{content});
            my $user_id = $user_data->{id};
            print "User ID: $user_id\n";

            my $profile_url = "https://node-community-api.teneo.pro/rest/v1/profiles?select=personal_code&id=eq.$user_id";
            my $profile_response = HTTP::Tiny->new->get($profile_url, {
                headers => {
                    'Authorization' => "Bearer $access_token",
                    'apikey' => $apikey
                }
            });

            if ($profile_response->{success}) {
                my $profile_data = decode_json($profile_response->{content});
                my $personal_code = $profile_data->[0]->{personal_code};
                print "Personal Code: $personal_code\n";
                set_user_id_to_file($user_id);
                set_account_data($email, $password, $access_token, $refresh_token, $personal_code);
                start_countdown_and_points();
                connect_websocket($user_id);
                print "$cl{gr}Data has been saved in the DataAccount.json file...\n$cl{rt}";
                CoderMark();
            }
        } else {
            print "Error: " . $response->{reason} . "\n";
        }
    }
}

sub reconnect_websocket {
    my $user_id = get_user_id_from_file();
    connect_websocket($user_id) if $user_id;
}

sub auto_login {
    my $account_data = get_account_data();
    if ($account_data->{email} && $account_data->{password}) {
        get_user_id();
        print "$cl{yl}\nAutomatic Login has been Successfully Executed..\n$cl{rt}";
    }
}

sub main {
    my $local_storage_data = get_local_storage();
    my $user_id = get_user_id_from_file();

    if (!$user_id) {
        my $option = prompt("\nUser ID not found. Would you like to:\n$cl{gr}\n1. Login to your account\n2. Enter User ID manually\n$cl{rt}\nChoose an option: ");
        if ($option == 1) {
            get_user_id();
        } elsif ($option == 2) {
            my $input_user_id = prompt('Please enter your user ID: ');
            set_user_id_to_file($input_user_id);
            start_countdown_and_points();
            connect_websocket($input_user_id);
        } else {
            print "Invalid option. Exiting...\n";
            exit(0);
        }
    } else {
        my $option = prompt("\nMenu:\n$cl{gr}\n1. Logout\n2. Start Running Node\n$cl{rt}\nChoose an option: ");
        if ($option == 1) {
            unlink 'UserId.json';
            unlink 'localStorage.json';
            unlink 'DataAccount.json';
            print "$cl{yl}\nLogged out successfully.\n";
            exit(0);
        } elsif ($option == 2) {
            CoderMark();
            print "$cl{gr}Initiates a connection to the node...\n$cl{rt}";
            start_countdown_and_points();
            connect_websocket($user_id);
        } else {
            print "$cl{yl}Invalid option. Exiting...\n";
            exit(0);
        }
    }
}

main();
AnyEvent->timer(0, 1800, \&auto_login);

File: GO/flynn.go
package main

import (
	"bufio"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"math"
	"math/rand"
	"net/http"
	"os"
	"os/signal"
	"strings"
	"syscall"
	"time"

	"github.com/gorilla/websocket"
)

var (
	socket            *websocket.Conn
	pingTicker        *time.Ticker
	countdownTicker   *time.Ticker
	potentialPoints   float64
	countdown         string
	pointsTotal       int
	pointsToday       int
	localStorageFile  = "localStorage.json"
	reader            = bufio.NewReader(os.Stdin)
)

type LocalStorage struct {
	LastUpdated     string  `json:"lastUpdated"`
	PointsTotal     int     `json:"pointsTotal"`
	PointsToday     int     `json:"pointsToday"`
	LastPingDate    string  `json:"lastPingDate"`
	PotentialPoints float64 `json:"potentialPoints"`
	Countdown       string  `json:"countdown"`
	UserId          string  `json:"userId"`
}

func getLocalStorage() (LocalStorage, error) {
	data, err := ioutil.ReadFile(localStorageFile)
	if err != nil {
		return LocalStorage{}, err
	}
	var localStorage LocalStorage
	err = json.Unmarshal(data, &localStorage)
	return localStorage, err
}

func setLocalStorage(data LocalStorage) error {
	currentData, _ := getLocalStorage()
	newData := currentData
	if data.LastUpdated != "" {
		newData.LastUpdated = data.LastUpdated
	}
	if data.PointsTotal != 0 {
		newData.PointsTotal = data.PointsTotal
	}
	if data.PointsToday != 0 {
		newData.PointsToday = data.PointsToday
	}
	if data.LastPingDate != "" {
		newData.LastPingDate = data.LastPingDate
	}
	if data.PotentialPoints != 0 {
		newData.PotentialPoints = data.PotentialPoints
	}
	if data.Countdown != "" {
		newData.Countdown = data.Countdown
	}
	if data.UserId != "" {
		newData.UserId = data.UserId
	}
	jsonData, err := json.Marshal(newData)
	if err != nil {
		return err
	}
	return ioutil.WriteFile(localStorageFile, jsonData, 0644)
}

func connectWebSocket(userId string) error {
	if socket != nil {
		return nil
	}
	version := "v0.2"
	url := "wss://secure.ws.teneo.pro"
	wsUrl := fmt.Sprintf("%s/websocket?userId=%s&version=%s", url, userId, version)
	var err error
	socket, _, err = websocket.DefaultDialer.Dial(wsUrl, nil)
	if err != nil {
		return err
	}

	connectionTime := time.Now().UTC().Format(time.RFC3339)
	err = setLocalStorage(LocalStorage{LastUpdated: connectionTime})
	if err != nil {
		return err
	}
	fmt.Println("WebSocket connected at", connectionTime)
	startPinging()
	startCountdownAndPoints()

	go func() {
		for {
			_, message, err := socket.ReadMessage()
			if err != nil {
				fmt.Println("WebSocket read error:", err)
				return
			}
			var data map[string]interface{}
			err = json.Unmarshal(message, &data)
			if err != nil {
				fmt.Println("JSON unmarshal error:", err)
				continue
			}
			fmt.Println("Received message from WebSocket:", data)
			if pointsTotal, ok := data["pointsTotal"].(float64); ok {
				if pointsToday, ok := data["pointsToday"].(float64); ok {
					lastUpdated := time.Now().UTC().Format(time.RFC3339)
					err = setLocalStorage(LocalStorage{
						LastUpdated: lastUpdated,
						PointsTotal: int(pointsTotal),
						PointsToday: int(pointsToday),
					})
					if err != nil {
						fmt.Println("Error setting local storage:", err)
					}
				}
			}
		}
	}()

	return nil
}

func disconnectWebSocket() {
	if socket != nil {
		socket.Close()
		socket = nil
		stopPinging()
	}
}

func startPinging() {
	stopPinging()
	pingTicker = time.NewTicker(10 * time.Second)
	go func() {
		for range pingTicker.C {
			if socket != nil {
				err := socket.WriteJSON(map[string]string{"type": "PING"})
				if err != nil {
					fmt.Println("Error sending ping:", err)
					continue
				}
				err = setLocalStorage(LocalStorage{LastPingDate: time.Now().UTC().Format(time.RFC3339)})
				if err != nil {
					fmt.Println("Error setting last ping date:", err)
				}
			}
		}
	}()
}

func stopPinging() {
	if pingTicker != nil {
		pingTicker.Stop()
		pingTicker = nil
	}
}

func startCountdownAndPoints() {
	if countdownTicker != nil {
		countdownTicker.Stop()
	}
	updateCountdownAndPoints()
	countdownTicker = time.NewTicker(1 * time.Second)
	go func() {
		for range countdownTicker.C {
			updateCountdownAndPoints()
		}
	}()
}

func updateCountdownAndPoints() {
	localStorage, err := getLocalStorage()
	if err != nil {
		fmt.Println("Error getting local storage:", err)
		return
	}

	if localStorage.LastUpdated != "" {
		lastUpdated, err := time.Parse(time.RFC3339, localStorage.LastUpdated)
		if err != nil {
			fmt.Println("Error parsing last updated time:", err)
			return
		}

		nextHeartbeat := lastUpdated.Add(15 * time.Minute)
		now := time.Now().UTC()
		diff := nextHeartbeat.Sub(now)

		if diff > 0 {
			minutes := int(diff.Minutes())
			seconds := int(diff.Seconds()) % 60
			countdown = fmt.Sprintf("%dm %ds", minutes, seconds)

			maxPoints := 25.0
			timeElapsed := now.Sub(lastUpdated)
			timeElapsedMinutes := timeElapsed.Minutes()
			newPoints := math.Min(maxPoints, (timeElapsedMinutes/15)*maxPoints)
			newPoints = math.Floor(newPoints*100) / 100

			if rand.Float64() < 0.1 {
				bonus := rand.Float64() * 2
				newPoints = math.Min(maxPoints, newPoints+bonus)
				newPoints = math.Floor(newPoints*100) / 100
			}

			potentialPoints = newPoints
		} else {
			countdown = "Calculating..."
			potentialPoints = 25
		}
	} else {
		countdown = "Calculating..."
		potentialPoints = 0
	}

	err = setLocalStorage(LocalStorage{
		PotentialPoints: potentialPoints,
		Countdown:       countdown,
	})
	if err != nil {
		fmt.Println("Error setting local storage:", err)
	}
}

func getUserId() error {
	loginUrl := "https://ikknngrgxuxgjhplbpey.supabase.co/auth/v1/token?grant_type=password"
	authorization := "Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6Imlra25uZ3JneHV4Z2pocGxicGV5Iiwicm9sZSI6ImFub24iLCJpYXQiOjE3MjU0MzgxNTAsImV4cCI6MjA0MTAxNDE1MH0.DRAvf8nH1ojnJBc3rD_Nw6t1AV8X_g6gmY_HByG2Mag"
	apikey := "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6Imlra25uZ3JneHV4Z2pocGxicGV5Iiwicm9sZSI6ImFub24iLCJpYXQiOjE3MjU0MzgxNTAsImV4cCI6MjA0MTAxNDE1MH0.DRAvf8nH1ojnJBc3rD_Nw6t1AV8X_g6gmY_HByG2Mag"

	fmt.Print("Email: ")
	email, _ := reader.ReadString('\n')
	email = strings.TrimSpace(email)

	fmt.Print("Password: ")
	password, _ := reader.ReadString('\n')
	password = strings.TrimSpace(password)

	payload := strings.NewReader(fmt.Sprintf(`{"email":"%s","password":"%s"}`, email, password))

	req, err := http.NewRequest("POST", loginUrl, payload)
	if err != nil {
		return err
	}

	req.Header.Add("Authorization", authorization)
	req.Header.Add("apikey", apikey)
	req.Header.Add("Content-Type", "application/json")

	res, err := http.DefaultClient.Do(req)
	if err != nil {
		return err
	}
	defer res.Body.Close()

	body, err := ioutil.ReadAll(res.Body)
	if err != nil {
		return err
	}

	var response map[string]interface{}
	err = json.Unmarshal(body, &response)
	if err != nil {
		return err
	}

	user, ok := response["user"].(map[string]interface{})
	if !ok {
		return fmt.Errorf("user data not found in response")
	}

	userId, ok := user["id"].(string)
	if !ok {
		return fmt.Errorf("user ID not found in response")
	}

	fmt.Println("User ID:", userId)

	profileUrl := fmt.Sprintf("https://ikknngrgxuxgjhplbpey.supabase.co/rest/v1/profiles?select=personal_code&id=eq.%s", userId)
	req, err = http.NewRequest("GET", profileUrl, nil)
	if err != nil {
		return err
	}

	req.Header.Add("Authorization", authorization)
	req.Header.Add("apikey", apikey)

	res, err = http.DefaultClient.Do(req)
	if err != nil {
		return err
	}
	defer res.Body.Close()

	body, err = ioutil.ReadAll(res.Body)
	if err != nil {
		return err
	}

	fmt.Println("Profile Data:", string(body))

	err = setLocalStorage(LocalStorage{UserId: userId})
	if err != nil {
		return err
	}

	startCountdownAndPoints()
	return connectWebSocket(userId)
}

func main() {
	localStorage, err := getLocalStorage()
	if err != nil {
		fmt.Println("Error getting local storage:", err)
	}

	userId := localStorage.UserId

	if userId == "" {
		fmt.Println("User ID not found. Would you like to:")
		fmt.Println("1. Login to your account")
		fmt.Println("2. Enter User ID manually")
		fmt.Print("Choose an option: ")
		option, _ := reader.ReadString('\n')
		option = strings.TrimSpace(option)

		switch option {
		case "1":
			err := getUserId()
			if err != nil {
				fmt.Println("Error getting user ID:", err)
				return
			}
		case "2":
			fmt.Print("Please enter your user ID: ")
			userId, _ = reader.ReadString('\n')
			userId = strings.TrimSpace(userId)
			err := setLocalStorage(LocalStorage{UserId: userId})
			if err != nil {
				fmt.Println("Error setting user ID:", err)
				return
			}
			startCountdownAndPoints()
			err = connectWebSocket(userId)
			if err != nil {
				fmt.Println("Error connecting to WebSocket:", err)
				return
			}
		default:
			fmt.Println("Invalid option. Exiting...")
			return
		}
	} else {
		fmt.Println("Menu:")
		fmt.Println("1. Logout")
		fmt.Println("2. Start Running Node")
		fmt.Print("Choose an option: ")
		option, _ := reader.ReadString('\n')
		option = strings.TrimSpace(option)

		switch option {
		case "1":
			err := os.Remove(localStorageFile)
			if err != nil {
				fmt.Println("Error removing local storage file:", err)
				return
			}
			fmt.Println("Logged out successfully.")
		case "2":
			startCountdownAndPoints()
			err := connectWebSocket(userId)
			if err != nil {
				fmt.Println("Error connecting to WebSocket:", err)
				return
			}
		default:
			fmt.Println("Invalid option. Exiting...")
			return
		}
	}

	// Handle SIGINT (Ctrl+C)
	c := make(chan os.Signal, 1)
	signal.Notify(c, os.Interrupt, syscall.SIGTERM)
	<-c

	fmt.Println("Received SIGINT. Stopping pinging...")
	stopPinging()
	disconnectWebSocket()
}

